source/server/listener_manager_impl.cc

```diff
@@ -490,19 +497,41 @@ const Network::FilterChain* ListenerImpl::findFilterChainForApplicationProtocols
   for (const auto& application_protocol : socket.requestedApplicationProtocols()) {
     const auto application_protocol_match = application_protocols_map.find(application_protocol);
     if (application_protocol_match != application_protocols_map.end()) {
-      return application_protocol_match->second.get();
+      return findFilterChainForSourceTypes(application_protocol_match->second, socket);
     }
   }
 
   // Match on a filter chain without application protocol requirements.
   const auto any_protocol_match = application_protocols_map.find(EMPTY_STRING);
   if (any_protocol_match != application_protocols_map.end()) {
-    return any_protocol_match->second.get();
+    return findFilterChainForSourceTypes(any_protocol_match->second, socket);
   }
 
   return nullptr;
 }
 
+const Network::FilterChain*
+ListenerImpl::findFilterChainForSourceTypes(const SourceTypesArray& source_types,
+                                            const Network::ConnectionSocket& socket) const {
+
+  auto filter_chain_local =
+      source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
+                       FilterChainMatch_ConnectionSourceType_LOCAL];
+
+  if (filter_chain_local && Network::Utility::isLocalConnection(socket)) {
+    return filter_chain_local.get();
+  }
+
+  auto filter_chain_external =
+      source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
+                       FilterChainMatch_ConnectionSourceType_EXTERNAL];
+  auto filter_chain_any =
+      source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
+                       FilterChainMatch_ConnectionSourceType_ANY];
+
+  return (filter_chain_external) ? filter_chain_external.get() : filter_chain_any.get();
```

Comment:
I believe the logic here is wrong. If someone only defines `source_type: EXTERNAL`, but the connection is local, you're still going to return filter chain for the external source type here.